### PR TITLE
bcachefs-tools: 1.4.1 -> 1.7.0

### PR DESCRIPTION
**Pull request by JohnRTitor**
## Description of changes

- Update: bcachefs-tools: 1.4.1 -> 1.7.0
- Changelog: https://github.com/koverstreet/bcachefs-tools/compare/v1.4.1...v1.7.0
- Adding myself as maintainer
- Reformat the file using `nixfmt-rfc-style`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---
Pinging maintainers: @davidak @Madouura 

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/filesystems/bcachefs-tools/default.nix [hunk 1]
```diff
@@ -43,14 +43,6 @@ stdenv.mkDerivation (finalAttrs: {
     makeWrapper
   ];
 
-  cargoRoot = "rust-src";
-  cargoDeps = rustPlatform.importCargoLock {
-    lockFile = ./Cargo.lock;
-    outputHashes = {
-      "bindgen-0.64.0" = "sha256-GNG8as33HLRYJGYe0nw6qBzq86aHiGonyynEM7gaEE4=";
```
[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: I tried building with `${src}/Cargo.lock`, didn't work as it would fail to find the file.

Removing `importCargoLock` will fail to build the rust code.

[the remark this answers is not in the document]
Comment (alyssais) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: We could use `cargoHash`.

[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: > We could use `cargoHash`.

Doesn't look like it's working.
```
┃        > unpacking source archive /nix/store/qyrbgxh64055v3v93dbasp8x6s2xhg4r-source
┃        > source root is source
┃        > Executing cargoSetupPostUnpackHook
┃        > unpacking source archive
┃        > do not know how to unpack source archive
```

[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: > I tried building with `${src}/Cargo.lock`, didn't work as it would fail to find the file.

Nevermind, I should have used `finalAttrs.src`. Successfully built. Pushing.

**Comment by JohnRTitor** [comment]
@ofborg eval
@ofborg build bcachefs-tools

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: Looks like ofborg eval is failing with latest changes. Successfully builds though.

[the remark this answers is not in the document]
Comment (alyssais) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: It's failing because it's import-from-derivation, which is not allowed in Nixpkgs.

> We are not using `rustPlatform.buildRustPackage` to build a "rust-only package" here.

Ah, then what we need is `rustPlatform.fetchCargoTarball`.

**Comment by JohnRTitor** [comment]
@ofborg build bcachefs-tools

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:50]: Nice, Ofborg eval working with fetchCargoTarball.
As for the update script, is there a script for automatically updating `hash` in `fetchCargoTarball`?

**Comment by JohnRTitor** [comment]
@Mic92 @alyssais is this ready to be approved? Latest bcachefs fixes and features also depend on `bcachefs-tools` package.

**Comment by alyssais** [comment]
Please squash the latest commit into the update, so there's not an intermediate point where it's using IFD.

**Comment by JohnRTitor** [comment]
@alyssais is this fine now? I decided to keep the last two commits separately, for obvious reasons.

File: pkgs/tools/filesystems/bcachefs-tools/default.nix [hunk 2]
```diff
@@ -65,20 +57,27 @@ stdenv.mkDerivation (finalAttrs: {
...
-  checkFlags = [ "BCACHEFS_TEST_USE_VALGRIND=no" ];
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    src = finalAttrs.src;
+    hash = "sha256-RsRz/nb8L+pL1U4l6RnvqeDFddPvcBFH4wdV7G60pxA=";
+  };
 
   makeFlags = [
-    "PREFIX=${placeholder "out"}"
+    "DESTDIR=${placeholder "out"}"
+    "PREFIX="
```
Comment (alyssais) [inline pkgs/tools/filesystems/bcachefs-tools/default.nix:67]: Are you sure this is right? https://github.com/NixOS/nixpkgs/issues/65718

File: pkgs/tools/filesystems/bcachefs-tools/default.nix [hunk 3]
```diff
@@ -65,20 +57,27 @@ stdenv.mkDerivation (finalAttrs: {
...
     "VERSION=${finalAttrs.version}"
     "INITRAMFS_DIR=${placeholder "out"}/etc/initramfs-tools"
-    "BCACHEFS_FUSE=${toString fuseSupport}"
-  ];
+  ] ++ lib.optional fuseSupport "BCACHEFS_FUSE=1";
```
Comment (alyssais) [inline pkgs/tools/filesystems/bcachefs-tools/default.nix:70]: Why this change?

File: pkgs/tools/filesystems/bcachefs-tools/default.nix [hunk 4]
```diff
@@ -65,20 +57,27 @@ stdenv.mkDerivation (finalAttrs: {
...
+
+
+  # FIXME: Try enabling this once the default linux kernel is at least 6.7
+  doCheck = false; # needs bcachefs module loaded on builder
 
   preCheck = lib.optionalString (!fuseSupport) ''
     rm tests/test_fuse.py
   '';
+  checkPhase = "./target/release/bcachefs version";
```
Comment (alyssais) [inline pkgs/tools/filesystems/bcachefs-tools/default.nix:79]: What's the point of setting this if `doCheck` is `false`?

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 3 above)
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:70]: IIRC, When fixing the build errors, I encountered "a libfuse3 not found error". I initially looked at bcachefs build.nix, to see what has changed leading to this. BCACHEFS_FUSE=0 had weird behaviour like trying to link unavailable library anyway.
With this change, this should fix the bug.

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 2 above)
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:67]: So a solution for this would be adding autotools to inputs and then using `installFlags = [ "localstatedir=${placeholder "out"}/var" ]; ` ?

I have not encountered any issues during normal usage with the new binary though.

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 4 above)
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:79]: I have added this check for future, when default LTS kernel becomes > 6.7. Then we can enable doCheck = true and that will run this script

Comment (alyssais) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:79]: But that overrides the default checkPhase, so now it won't run any of the tests it would have run before.

What's the motivation for testing that `bcachefs version` works?  It feels like not a very meaningful test unless I'm missing something.

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 2 above)
Comment (alyssais) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:67]: No, because this package doesn't use autotools, and localstatedir would only cover that one directory, not the whole thing.

The right thing to do is generally set PREFIX, and not set DESTDIR, like it was before.  Why did you change it?

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 4 above)
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:79]: Oh, I thought I would add a step to checkPhase not override it. While `bcachefs --version` is not an extensive check, it could show if there is a fatal runtime error due to missing libraries/packages.

I'll revert this so default checkPhase is used, if enabled that is.

File: pkgs/tools/filesystems/bcachefs-tools/default.nix (hunk 2 above)
Comment (JohnRTitor) [reply pkgs/tools/filesystems/bcachefs-tools/default.nix:67]: Looks like it was a oversight on my part. I was trying to cleanup this config by looking at the build.nix from the repository. It modify any existing behaviour at first glance, so I thought it was okay to include the change.

I'll revert this change.

**Comment by alyssais** [comment]
Test failure on x86_64-linux should probably be investigated.  Is it just racy perhaps?

**Comment by JohnRTitor** [comment]
> Test failure on x86_64-linux should probably be investigated. Is it just racy perhaps?

When I ran the test locally it did not actually "fail", it tested everything including in a VM, but then it took a lot of time at the finishing stage, [leading to the timeout](https://github.com/NixOS/nixpkgs/pull/304761/checks?check_run_id=24206442471).

**Comment by JohnRTitor** [comment]
@Mic92 @alyssais do I need to make any other changes or is this ready to be merged?

**Review by onny — APPROVED** [review]

**Comment by onny** [comment]
Result of `nixpkgs-review pr 304761` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bcachefs-tools</li>
  </ul>
</details>

**Review by alyssais — APPROVED** [review]

**Comment by onny** [comment]
Thank you!
